### PR TITLE
Friendships-v1

### PR DESCRIPTION
**Pull request by elpotatoslayer**
In this Milestone I :
-- created methods to accept, reject, and many more inside the user model
-- Implement friendship controller
-- Implement an 'Invite to Friendship' link through which a user can send friend requests to other users,
-- A user can either Accept or Reject the friendship request of others.
-- Wrote unit tests for models and integrations tests

### CODE REVIEW

File: app/views/users/_singlepage.html.erb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    
+    <%= render 'friendships/reject', user: @user, action_text: 'Cancel request', method: 'delete' %>
+  <% elsif current_user.friend_requests.include?(@user) %>
+    
+    <%= render 'friendships/accept_friend', user: @user, action_text: 'Accept friend', method: 'put' %>
+    <%= render 'friendships/reject', user: @user, action_text: 'Reject', method: 'delete' %>
+  <% else %>
+    <%= render 'friendships/add_friend', user: @user, action_text: 'Add friend', method: 'post' %>
+  <% end %>
+     <% end %>
```
Comment (mahmoud717) [inline app/views/users/_singlepage.html.erb:14]: - [ ] please make sure to add this logic to the helpers.

File: app/views/users/_user.html.erb [hunk 2]
```diff
@@ -2,5 +2,20 @@
...
+    
+        <%= render 'friendships/reject', user: user, action_text: 'Cancel request', method: 'delete' %>
+      <% elsif current_user.friend_requests.include?(user) %>
+    
+        <%= render 'friendships/accept_friend', user: user, action_text: 'Accept friend', method: 'put' %>
+        <%= render 'friendships/reject', user: user, action_text: 'Reject', method: 'delete' %>
+      <% else %>
+        <%= render 'friendships/add_friend', user: user, action_text: 'Add friend', method: 'post' %>
+      <% end %>
+    <% end %>
```
Comment (mahmoud717) [inline app/views/users/_user.html.erb:19]: - [ ] Please make sure to add this logic to the helpers.

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @Ralph-1**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [ ] Please make sure to test all associations and validations and to add an integration test.
- [ ] When User1 sends an invite to User2 and User2 accepts the invite, there is an issue that only User1 can see his own posts and friends' posts, while User2 only sees his own posts.
you should make sure that both users see each other's posts after they become friends.

- [ ] Please make sure to remove all the logic from the views, and by logic I mean if statements, unless statements, and loops and add that to the helpers.



---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Ralph Oburu** 
Good job with the improvements for this project :muscle:  :muscle:
It meets most of the requirements :clap: :clap: :clap:  
There are just a few details to improve before you can go ahead:

- [ ] Remove commented code in spec/models/user_spec.rb
- [ ] Add some tests in spec/models/comment_spec.rb
- [ ] Add integrations tests
- [ ] After logging, the root page of the website should only show comments from the current user and its friends.
- [ ] Cancel request text isn't fully visible


 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by MahmoudBakr23 — CHANGES_REQUESTED** [review]
# **Changes Required 🛑♻**

<hr>

Hello there :)
Thanks for the extra features you made. It's really cool 💯 
Great work and effort you put into this milestone. Hang on! You're almost there :) 👍 
There are still a few more changes left to be good to go.

<hr>

- [ ] Please make sure to fix the logic for the timeline posts. Remember that a user cannot see the posts that belong to another user unless both users are friends already!
- [ ] As requested earlier, there should be no logic in the views. Please refer to the first review and read it more carefully.
- **OPTIONAL** I'm sorry that this requirement didn't brought up earlier, but both "Add Friend" and "Accept, Reject" buttons should appear on both "All Users" and each user's profile pages.

<hr>

That's it :)
Good luck and happy coding 😊👨‍💻💻

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @Ralph-1**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [ ] You were requested to move your logical code to a helper, not a partial file. Partial files generally contain forms and should not contain logical code. Please make sure to remove all of your logic to a "HELPER" file. 

---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi  @Ralph-1 
Good job with the changes! :tada: We are almost there.
We, however, do have a few changes to carry out before we proceed.

##

 For this project, we have to make sure nothing is messed up. Kindly compare the screenshots below 👇 and work on the suggested changes.
- [ ] When we save the post it is duplicating on the page. For better understanding please watch the [video](https://www.loom.com/share/2e5d15054e1c4e4f8488f60fe07df6a3). 
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

File: app/views/users/_user.html.erb [hunk 3]
```diff
@@ -2,5 +2,20 @@
...
+    
+        <%= render 'friendships/reject', user: user, action_text: 'Cancel request', method: 'delete' %>
+      <% elsif current_user.friend_requests.include?(user) %>
+    
+        <%= render 'friendships/accept_friend', user: user, action_text: 'Accept friend', method: 'put' %>
+        <%= render 'friendships/reject', user: user, action_text: 'Reject', method: 'delete' %>
+      <% else %>
+        <%= render 'friendships/add_friend', user: user, action_text: 'Add friend', method: 'post' %>
+      <% end %>
+     <% end %>
```
Comment (acushlakoncept) [inline app/views/users/_user.html.erb:19]: There should be absolutely NO logic in views.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

Hello @Ralph-1, 

You've done an awesome job so far 👌. But there are still some changes that are necessary for this project to be approved.

- [ ] You still have logic in your views. There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
I have gone ahead and made some comments on your codebase see affected files below:

Also inside your `layout/application.html.erb`
```ruby
<% if current_user %>
        <%= link_to 'Sign out', destroy_user_session_path, method: :delete %>
      <% else %>
        <%= link_to 'Sign in', user_session_path %>
      <% end %>
    </nav>
    <% if notice.present? %>
      <div class="notice">
        <p> <%= notice %> </p>
      </div>
    <% end %>
    <% if alert.present? %>
      <div class="alert">
        <p> <%= alert %> </p>
      </div>
    <% end %>
```

**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

- [ ] This will be considered OPTIONAL since the previous TSEs may have missed it. But consider it important feature to implement as it will help you while moving to the next milestone. 
* Posts timeline should include posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)). Currently all posts are been displayed on the timeline.
* I see you have nicely implemented unit tests on all your model, kindly add integration tests as well.  [Capybara](https://github.com/teamcapybara/capybara) is an exceptionally great tool for performing an integration test with RSpec because it helps you perform end-to-end tests on your applications. It does this by simulating how a real user would interact with your application. Want to know more? See this [article](https://www.microverse.org/blog/understanding-rspec-controller-request-specs-integration-test-using-capybara).

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done lads and thanks for implementing the changes requested earlier on**. However, there are a few issues to fix to get approved.

### REQUIRED CHANGES: ⚠️ 

- [ ] You still have some logic statements in the `views/layouts/application.html.erb` file. Move them to the helpers directory just the same way you did with the other logic statements

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

## Ping me on slack for a chit-chat 🆘  Roy Mukuye Or mention me in this PR

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:
To Highlight: :angel:

- [ ] User can invite friends :+1: 
- [ ] User can see friendships requests :100: 
- [ ] User has to accept friendship request to start friendship :1st_place_medal: 
- [ ] User can reject a friendship request :+1: 
- [ ] There is a button to invite user to friendship on users list page :100: 
- [ ] Posts timeline includes posts written by logged-in user and their friends  :1st_place_medal: 

To Improve :factory_worker:

- [ ] Please, make sure that you test are not failing because, it means that there is bug in your APP:

In fact, software testing is the process of verifying and validating a software application to check whether it is working as expected.

- [ ] It will be perfect if you remove also logic in your partial files and put it in your helpers. As a developer your job is to create quality code that solves your client’s problems. Quality has many aspects to it, including readability and maintainability. It is much easier to read code that is lean and skinny, as well as serving a single purpose. When you look at a CSS file, you know it’s for styling and giving your site personality and uniqueness. When you look at a JavaScript file, you know it handles the interactivity on the front-end for the person viewing the site. And the view file is responsible for the HTML. (_user.html.erb)

- [ ] It will be nice if you add a button to invite user to friendship on single user page  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

**Comment by elpotatoslayer** [comment]
@rindrajosia run the project again, please. Cause all the changes that you are requesting already exists. Kindly rerun the project and you will see that the project meets all the requirements.

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @Ralph-1  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, this milestone is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [ ] Tests are failing, with the following error message:
```ruby
Failures:

  1) User associations is expected to have many friends_posts through friends source => posts
     Failure/Error: it { is_expected.to have_many(:friends_posts).through(:friends).source(:posts) }
       Expected User to have a has_many association called friends_posts through friends (Cannot have a has_many :through association 'User#friends_posts' which goes through 'User#friends' before the through association is defined.)
     # ./spec/models/user_spec.rb:15:in `block (3 levels) in <main>'

Finished in 0.23407 seconds (files took 0.74479 seconds to load)
19 examples, 1 failure

Failed examples:

rspec ./spec/models/user_spec.rb:15 # User associations is expected to have many friends_posts through friends source => posts
```
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Review by andresporras3423 — APPROVED** [review]
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:
**Hello Ralph Oburu** 
Amazing job :tada: :tada: :tada:  you have met all the requirements for this project
Now you are ready to continue with your next challenge :muscle: 
Don't forget to merge with the develop branch
 
Keep going like this  :clap: :clap:
 
Oscar Andrés Russi Porras

**Pull request merged**